### PR TITLE
Add metadata ADR

### PR DESCRIPTION
**Pull request by Jarema**
# Metadata for Stream and Consumer

|Metadata|Value|
|--------|-----|
|Date    |2023-01-23|
|Author  |@Jarema|
|Tags    |jetstream, client, server|

## Context and Problem Statement

Until now, there was no way to easily add additional information about Stream or Consumer.
The only solution was using `Description` field, which is a ugly workaround.

## Server PR
https://github.com/nats-io/nats-server/pull/3797

## Design

The solution is to add new `metadata` field to both `Consumer` and `Stream` config.
The `metadata` field would by a map of `string` keys and `string` values.

The map would be represented in json as object with nested key/value pairs, which is a default
way to marshal maps/hashmaps in most languages.

### Example
```json
{
  "durable_name": "consumer",
  ... // other consumer/stream fields
  "metadata": {
    "owner": "nack",
    "domain": "product"
  }
}

```

## Open questions

Do we want to limit the size of `metadata`?

Signed-off-by: Tomasz Pietrek <tomasz@nats.io>

### CODE REVIEW

File: adr/ADR-00.md
Comment (ripienaar) [inline adr/ADR-00.md:26]: 1MB is very big - given nats messages are capped at 1MB, I'd aim for way lower.  1MB string:string map is huge

Comment (Jarema) [reply adr/ADR-00.md:26]: I'm fine with clamping to 512kB or even 128kB.

@aricart any thoughts here?

Comment (Jarema) [reply adr/ADR-00.md:26]: @wallyqs I guess even 128kB would be more than enough for anything orchestration related, like k8s.

Comment (ripienaar) [reply adr/ADR-00.md:26]: Afaik these are k8s label limits for comparison ```Each resource can have multiple cluster labels, up to a maximum of 64. Each cluster label must be a key-value pair. Keys have a minimum length of 1 character and a maximum length of 63 characters, and cannot be empty. Values can be empty, and have a maximum length of 63 characters.```

Comment (wallyqs) [reply adr/ADR-00.md:26]: 128kb sounds more than enough, similar limits as k8s works too.

File: adr/ADR-00.md
Comment (Jarema) [reply adr/ADR-00.md:36]: The internal namespace is a good idea!

Maybe `_nats` ?

Those could be use by both client internals and the server.
Server could ensure immutability of its entries.

Comment (wallyqs) [reply adr/ADR-00.md:36]: `_nats` sounds good 👍

File: adr/ADR-00.md
Comment (Jarema) [reply adr/ADR-00.md:26]: Ok, 128k it is.

**Comment by Jarema** [comment]
Improved ADR after discussions:

1. Set limit to 128kB
2. Added reserved prefix for internals `_nats`.

File: adr/ADR-00.md
Comment (aricart) [reply adr/ADR-00.md]: 128K sounds good to me as well

File: adr/ADR-00.md
Comment (ripienaar) [inline adr/ADR-00.md:7]: Make this `Approved`, once merged it will be approved and we know we already have a PR in the server, Proposed only really for stuff thats just exploring ideas and with no immediate hopes for implementation

File: adr/ADR-00.md
Comment (ripienaar) [inline adr/ADR-00.md:45]: Be great to come up with a list of these.

File: adr/ADR-00.md
Comment (ripienaar) [inline adr/ADR-00.md:34]: You mean the server would reject updating labels that start with `_nats`? Not sure that as a broad rule is a good one

Comment (Jarema) [reply adr/ADR-00.md:34]: No no, I just refer @wallyqs idea to have some fields possibly immutable - like those set by the server, not the clients - with what server version consumer/stream was created etc.
So per field basis - when we introduce those.

_nats should be only internal and clients should  probably validate that.

Comment (ripienaar) [reply adr/ADR-00.md:34]: Ah I see, yeah all good then. We should make an effort to recommend a few that clients can all agree on. Probably not in this PR though.

**Comment by Jarema** [comment]
@ripienaar @wallyqs  ADR updated, PR updated.

**Review by wallyqs — APPROVED** [review]

**Review by ripienaar — APPROVED** [review]
LGTM

**Pull request merged**
